contracts/enclaves/AttestationAutherUpgradeable.sol

```diff
@@ -0,0 +1,156 @@
+import "@openzeppelin/contracts-upgradeable/access/AccessControlEnumerableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import "./interfaces/IAttestationVerifier.sol";
+
+contract AttestationAutherUpgradeable is
+    Initializable  // initializer
+{
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    IAttestationVerifier public immutable ATTESTATION_VERIFIER;
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    uint256 public immutable ATTESTATION_MAX_AGE;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(IAttestationVerifier attestationVerifier, uint256 maxAge) {
+        ATTESTATION_VERIFIER = attestationVerifier;
+        ATTESTATION_MAX_AGE = maxAge;
+    }
+
+    struct EnclaveImage {
+        bytes PCR0;
+        bytes PCR1;
+        bytes PCR2;
+    }
+
+    mapping(bytes32 => EnclaveImage) public whitelistedImages;
+    mapping(address => bytes32) public verifiedKeys;
+
+    uint256[48] private __gap;
+
+    event EnclaveImageWhitelisted(bytes32 indexed imageId, bytes PCR0, bytes PCR1, bytes PCR2);
+    event EnclaveImageRevoked(bytes32 indexed imageId);
+    event EnclaveKeyWhitelisted(bytes indexed enclavePubKey, bytes32 indexed imageId);
+    event EnclaveKeyRevoked(bytes indexed enclavePubKey);
+    event EnclaveKeyVerified(bytes indexed enclavePubKey, bytes32 indexed imageId);
+
+    function __AttestationAuther_init_unchained(EnclaveImage[] memory images) internal onlyInitializing {
+        for (uint256 i = 0; i < images.length; i++) {
+            _whitelistEnclaveImage(images[i]);
+        }
+    }
+
+    function _pubKeyToAddress(bytes memory pubKey) internal pure returns (address) {
+        require(pubKey.length == 64, "Invalid public key length");
+
+        bytes32 hash = keccak256(pubKey);
+        return address(uint160(uint256(hash)));
+    }
+
+    function _whitelistEnclaveImage(EnclaveImage memory image) internal returns(bytes32) {
+        require(
+            image.PCR0.length == 48 &&
+            image.PCR1.length == 48 &&
+            image.PCR2.length == 48,
+            "AA:WI-PCR values must be 48 bytes"
+        );
+
+        bytes32 imageId = keccak256(abi.encodePacked(image.PCR0, image.PCR1, image.PCR2));
+        require(whitelistedImages[imageId].PCR0.length == 0, "AA:WI-image already whitelisted");
+        whitelistedImages[imageId] = EnclaveImage(image.PCR0, image.PCR1, image.PCR2);
+        emit EnclaveImageWhitelisted(imageId, image.PCR0, image.PCR1, image.PCR2);
+        return imageId;
+    }
+
+    function _revokeEnclaveImage(bytes32 imageId) internal {
+        require(whitelistedImages[imageId].PCR0.length != 0, "AA:RI-Image not whitelisted");
+
+        delete whitelistedImages[imageId];
+        emit EnclaveImageRevoked(imageId);
+    }
+
+    function _whitelistEnclaveKey(bytes memory enclavePubKey, bytes32 imageId) internal {
```

Comment:
Should this even exist? Ideally we don't want people to randomly whitelist enclave keys na

Reply:
it's internal, in case ppl want to add admin functions

Reply:
Inclined to leave it just so it is a complete set, but yeah, probably shouldn't really be used

Reply:
Ahh... okay. We can probably remove `whitelistEnclaveKey` and keep the internal methods?

Reply:
it is internal, external is only in the sample